edu-services/cm-service/cm-impl/hibernate-impl/impl/src/java/org/sakaiproject/coursemanagement/impl/CourseSiteRemovalServiceImpl.java

```diff
@@ -0,0 +1,218 @@
+      this.securityService = securityService;
+   }
+
+   /**
+    * returns the instance of the SiteService injected by the spring framework specified in the components.xml file via IoC.
+    * <br/><br/>
+    * @return the instance of the SiteService injected by the spring framework specified in the components.xml file via IoC.
+    */
+   public SiteService getSiteService() {
+      return siteService;
+   }
+
+   /**
+    * called by the spring framework to initialize the siteService data member specified in the components.xml file via IoC.
+    * <br/><br/>
+    * @param siteService   the implementation of the SiteService interface provided by the spring framework.
+    */
+   public void setSiteService(SiteService siteService) {
+      this.siteService = siteService;
+   }
+
+   /**
+    * removes\\unpublishes course sites whose terms have ended and a specified number of days have passed.
+    * Once a term has ended, the course sites for that term remain available for a specified number of days, whose duration is specified in sakai.properties
+    * via the <i>course_site_removal_service.num_days_after_term_ends</i> property.  After the specified period has elapsed, this invoking this service will either
+    * remove or unpublish the course site, depending on the value of the <i>course_site_removal_service.action</i> sakai property.
+    * </br></br>
+    * @param action                 whether to delete the course site or to simply unpublish it.
+    * @param numDaysAfterTermEnds   number of days after a term ends when course sites expire.
+    * </br></br>
+    * @return the number of course sites that were removed\\unpublished.
+    */
+
+    public int removeCourseSites(CourseSiteRemovalService.Action action, int numDaysAfterTermEnds) {
+
+       logger.info("removeCourseSites(" + action + " course sites, " + numDaysAfterTermEnds + " days after the term ends)");
+       Date today           = new Date();
+       Date expirationDate  = new Date(today.getTime() - numDaysAfterTermEnds * ONE_DAY_IN_MS);
+       int  numSitesRemoved = 0;
+
+       try {
+         // get the list of the academic term(s)
+         List<AcademicSession> academicSessions = courseManagementService.getAcademicSessions();
+
+         for(AcademicSession academicSession : academicSessions) {
+            // see if the academic session ended more than the specified number of days ago
+            if (academicSession.getEndDate().getTime() < expirationDate.getTime()) {
+               // get a list of all published course sites in ascending creation date order which are associated with the specified academic session
+               Hashtable<String, String> propertyCriteria = new Hashtable<String, String>();
+               propertyCriteria.put("term_eid", academicSession.getEid());
+                List<String> sites = (List<String>)siteService.getSiteIds(SelectionType.PUBVIEW, "course", null, propertyCriteria, SortType.CREATED_ON_ASC, null);
```

Comment:
Sorry to comment on an old PR .... but I am just now testing this code......

Two issues I see with this call to siteService.getSiteIds:

1) The "admin" user must be enrolled in the site for it to be picked up because of this code in DbSiteService:

```
            if ((type == SelectionType.ACCESS) || (type == SelectionType.UPDATE) || (type == SelectionType.MEMBER) || (type == SelectionType.DELETED) || (type == SelectionType.PUBVIEW) || (type == SelectionType.INACTIVE_ONLY))            {
                // join with the SITE_USER table
                join = siteServiceSql.getSitesJoin1Sql();
            }   
```    

2) The query only asks for sites that have PUBVIEW=1 .... I would think that PUBVIEW would be irrelevant